File: Java Playground/07_BinarySearch.java
import java.util.Arrays;

/**
 * Binary Search Template & Variants.
 *
 * <p><b>CORE INVARIANT:</b> Maintain arr[left..mid-1] < target ≤ arr[mid..right]
 * (or similar depending on search type).
 *
 * <p><b>LOOP CONDITION:</b>
 * - Use `left < right` when searching for a boundary (first/last occurrence).
 * - Use `left <= right` when searching for exact match and returning -1 if not found.
 *
 * <p><b>MID CALCULATION:</b> Always use `mid = left + (right - left) / 2` to avoid
 * integer overflow. NEVER use `(left + right) / 2`.
 *
 * <p><b>BINARY SEARCH ON ANSWER:</b> Instead of searching array, define feasibility
 * function and binary search on answer space (e.g., min eating speed, min allocation).
 *
 * <p>Time: O(log n) for standard, O(log n * check) for answer space. Space: O(1).
 */
public class BinarySearch {

  /**
   * Standard Binary Search - Find exact match.
   *
   * @param nums sorted array
   * @param target value to find
   * @return index of target if found, -1 otherwise
   * @throws NullPointerException if nums is null
   *
   *         <p>LeetCode 704 (Easy). Time: O(log n). Space: O(1).
   *
   *         <p>Use `left <= right` for exact match. Return immediately when
   *         found.
   */
  public static int search(int[] nums, int target) {
    if (nums == null) {
      return -1;
    }

    int left = 0;
    int right = nums.length - 1;

    while (left <= right) {
      int mid = left + (right - left) / 2;

      if (nums[mid] == target) {
        return mid;
      } else if (nums[mid] < target) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }

    return -1;
  }

  /**
   * Search Insert Position.
   *
   * @param nums sorted array
   * @param target value to find or insert
   * @return index where target is found, or where it should be inserted to
   *     maintain sorted order
   * @throws NullPointerException if nums is null
   *
   *         <p>LeetCode 35 (Easy). Time: O(log n). Space: O(1).
   *
   *         <p>Use `left < right` to find left boundary. At end, left is
   *         insertion position.
   */
  public static int searchInsertPosition(int[] nums, int target) {
    if (nums == null) {
      return 0;
    }

    int left = 0;
    int right = nums.length;

    while (left < right) {
      int mid = left + (right - left) / 2;

      if (nums[mid] < target) {
        left = mid + 1;
      } else {
        right = mid;
      }
    }

    return left;
  }

  /**
   * Find First Occurrence of target (smallest index).
   *
   * @param nums sorted array (may have duplicates)
   * @param target value to find
   * @return smallest index where nums[i] == target, or -1 if not found
   * @throws NullPointerException if nums is null
   *
   *         <p>Time: O(log n). Space: O(1).
   *
   *         <p>Use `left < right` to find boundary. When nums[mid] == target,
   *         move right = mid to keep searching left. At end, check if nums[left]
   *         == target.
   */
  public static int findFirstOccurrence(int[] nums, int target) {
    if (nums == null || nums.length == 0) {
      return -1;
    }

    int left = 0;
    int right = nums.length;

    while (left < right) {
      int mid = left + (right - left) / 2;

      if (nums[mid] < target) {
        left = mid + 1;
      } else {
        right = mid;
      }
    }

    return left < nums.length && nums[left] == target ? left : -1;
  }

  /**
   * Find Last Occurrence of target (largest index).
   *
   * @param nums sorted array (may have duplicates)
   * @param target value to find
   * @return largest index where nums[i] == target, or -1 if not found
   * @throws NullPointerException if nums is null
   *
   *         <p>Time: O(log n). Space: O(1).
   *
   *         <p>Use `left < right` to find boundary. When nums[mid] == target,
   *         move left = mid + 1 to keep searching right. At end, check if
   *         nums[right - 1] == target.
   */
  public static int findLastOccurrence(int[] nums, int target) {
    if (nums == null || nums.length == 0) {
      return -1;
    }

    int left = 0;
    int right = nums.length;

    while (left < right) {
      int mid = left + (right - left) / 2;

      if (nums[mid] <= target) {
        left = mid + 1;
      } else {
        right = mid;
      }
    }

    return left > 0 && nums[left - 1] == target ? left - 1 : -1;
  }

  /**
   * Search Range - Find first and last occurrence in one call.
   *
   * @param nums sorted array
   * @param target value to find
   * @return int[] {first, last} indices, or {-1, -1} if not found
   * @throws NullPointerException if nums is null
   *
   *         <p>LeetCode 34 (Medium). Time: O(log n). Space: O(1).
   *
   *         <p>Call findFirstOccurrence and findLastOccurrence helper functions.
   */
  public static int[] searchRange(int[] nums, int target) {
    if (nums == null || nums.length == 0) {
      return new int[]{-1, -1};
    }

    int first = findFirstOccurrence(nums, target);
    if (first == -1) {
      return new int[]{-1, -1};
    }

    int last = findLastOccurrence(nums, target);
    return new int[]{first, last};
  }

  /**
   * Search in Rotated Sorted Array (unknown pivot).
   *
   * @param nums rotated sorted array with distinct elements
   * @param target value to find
   * @return index of target, or -1 if not found
   * @throws NullPointerException if nums is null
   *
   *         <p>LeetCode 33 (Medium). Time: O(log n). Space: O(1).
   *
   *         <p>Determine which half is sorted, check if target in sorted half.
   *         If yes, search that half; else search other half.
   */
  public static int searchInRotatedArray(int[] nums, int target) {
    if (nums == null || nums.length == 0) {
      return -1;
    }

    int left = 0;
    int right = nums.length - 1;

    while (left <= right) {
      int mid = left + (right - left) / 2;

      if (nums[mid] == target) {
        return mid;
      }

      // Determine which half is sorted
      if (nums[left] <= nums[mid]) {
        // Left half is sorted
        if (nums[left] <= target && target < nums[mid]) {
          // Target in sorted left half
          right = mid - 1;
        } else {
          left = mid + 1;
        }
      } else {
        // Right half is sorted
        if (nums[mid] < target && target <= nums[right]) {
          // Target in sorted right half
          left = mid + 1;
        } else {
          right = mid - 1;
        }
      }
    }

    return -1;
  }

  /**
   * Find Minimum in Rotated Sorted Array.
   *
   * @param nums rotated sorted array with distinct elements
   * @return minimum element
   * @throws NullPointerException if nums is null
   *
   *         <p>LeetCode 153 (Medium). Time: O(log n). Space: O(1).
   *
   *         <p>Rotation point where min occurs. If nums[mid] > nums[right],
   *         min is in right half. If nums[mid] <= nums[right], min is in left
   *         half (including mid).
   */
  public static int findMinInRotatedArray(int[] nums) {
    if (nums == null || nums.length == 0) {
      return Integer.MIN_VALUE;
    }

    int left = 0;
    int right = nums.length - 1;

    // Already sorted (no rotation)
    if (nums[left] <= nums[right]) {
      return nums[left];
    }

    while (left < right) {
      int mid = left + (right - left) / 2;

      if (nums[mid] > nums[right]) {
        // Minimum in right half
        left = mid + 1;
      } else {
        // Minimum in left half or at mid
        right = mid;
      }
    }

    return nums[left];
  }

  /**
   * Koko Eating Bananas (binary search on answer space).
   *
   * @param piles array of pile sizes
   * @param h hours available
   * @return minimum eating speed (bananas per hour) to finish all piles
   * @throws NullPointerException if piles is null
   *
   *         <p>LeetCode 875 (Medium). Time: O(n log m) where m = max pile size.
   *         Space: O(1).
   *
   *         <p>Search on answer space [1, max(piles)]. For each speed, check if
   *         Koko can finish in h hours.
   */
  public static int kokoEatingBananas(int[] piles, int h) {
    if (piles == null || piles.length == 0) {
      return 0;
    }

    int left = 1;
    int right = 0;
    for (int pile : piles) {
      right = Math.max(right, pile);
    }

    while (left < right) {
      int mid = left + (right - left) / 2;

      if (canFinish(piles, mid, h)) {
        right = mid;
      } else {
        left = mid + 1;
      }
    }

    return left;
  }

  /**
   * Helper: can Koko eat all piles at given speed in h hours?
   *
   * @param piles array of pile sizes
   * @param speed eating speed (bananas/hour)
   * @param h available hours
   * @return true if can finish all piles
   */
  private static boolean canFinish(int[] piles, int speed, int h) {
    long hours = 0;
    for (int pile : piles) {
      hours += (pile + speed - 1) / speed; // Ceiling division
    }
    return hours <= h;
  }

  /**
   * Split Array Largest Sum (binary search on answer space).
   *
   * @param nums array to split
   * @param k number of subarrays to split into
   * @return minimized largest sum among all k subarrays
   * @throws NullPointerException if nums is null
   *
   *         <p>LeetCode 410 (Hard). Time: O(n log(sum)).
   *         Space: O(1).
   *
   *         <p>Binary search on answer: largest possible sum. For each sum,
   *         check if can split into k subarrays where each sum ≤ answer.
   */
  public static int splitArrayLargestSum(int[] nums, int k) {
    if (nums == null || nums.length == 0) {
      return 0;
    }

    int left = 0;
    int right = 0;

    for (int num : nums) {
      left = Math.max(left, num);  // At least largest element
      right += num;                // At most total sum
    }

    while (left < right) {
      int mid = left + (right - left) / 2;

      if (canSplit(nums, mid, k)) {
        right = mid;
      } else {
        left = mid + 1;
      }
    }

    return left;
  }

  /**
   * Helper: can split array into k subarrays where each sum ≤ maxSum?
   *
   * @param nums input array
   * @param maxSum maximum allowed sum per subarray
   * @param k number of subarrays
   * @return true if possible
   */
  private static boolean canSplit(int[] nums, int maxSum, int k) {
    int subarrays = 1;
    long currentSum = 0;

    for (int num : nums) {
      if (currentSum + num > maxSum) {
        subarrays++;
        currentSum = num;

        if (subarrays > k) {
          return false;
        }
      } else {
        currentSum += num;
      }
    }

    return true;
  }

  /**
   * Ship Packages Within Days (binary search on answer space).
   *
   * @param weights array of package weights
   * @param days number of days to ship all packages
   * @return minimum capacity (weight per day) required
   * @throws NullPointerException if weights is null
   *
   *         <p>LeetCode 1011 (Medium). Time: O(n log(sum)).
   *         Space: O(1).
   *
   *         <p>Binary search on capacity [max weight, total weight]. Check if
   *         given capacity allows shipping in days.
   */
  public static int shipPackagesWithinDays(int[] weights, int days) {
    if (weights == null || weights.length == 0) {
      return 0;
    }

    int left = 0;
    int right = 0;

    for (int w : weights) {
      left = Math.max(left, w);
      right += w;
    }

    while (left < right) {
      int mid = left + (right - left) / 2;

      if (canShip(weights, mid, days)) {
        right = mid;
      } else {
        left = mid + 1;
      }
    }

    return left;
  }

  /**
   * Helper: can ship all packages in days with given capacity?
   *
   * @param weights array of weights
   * @param capacity capacity per day
   * @param days days available
   * @return true if possible
   */
  private static boolean canShip(int[] weights, int capacity, int days) {
    int daysNeeded = 1;
    long currentWeight = 0;

    for (int w : weights) {
      if (currentWeight + w > capacity) {
        daysNeeded++;
        currentWeight = w;

        if (daysNeeded > days) {
          return false;
        }
      } else {
        currentWeight += w;
      }
    }

    return true;
  }

  /**
   * Search Matrix (2D, treated as 1D sorted array).
   *
   * @param matrix 2D sorted matrix (row-wise and column-wise)
   * @param target value to find
   * @return true if target found, false otherwise
   * @throws NullPointerException if matrix is null
   *
   *         <p>LeetCode 74 (Medium). Time: O(log(m*n)).
   *         Space: O(1).
   *
   *         <p>Treat 2D matrix as flattened 1D array. mid -> [mid/cols][mid%cols].
   */
  public static boolean searchMatrix(int[][] matrix, int target) {
    if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
      return false;
    }

    int rows = matrix.length;
    int cols = matrix[0].length;
    int left = 0;
    int right = rows * cols - 1;

    while (left <= right) {
      int mid = left + (right - left) / 2;
      int val = matrix[mid / cols][mid % cols];

      if (val == target) {
        return true;
      } else if (val < target) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }

    return false;
  }

  /**
   * Search Matrix II (2D, row and column sorted but no overall ordering).
   *
   * @param matrix 2D matrix where each row and column is sorted
   * @param target value to find
   * @return true if target found, false otherwise
   * @throws NullPointerException if matrix is null
   *
   *         <p>LeetCode 240 (Medium). Time: O(m + n).
   *         Space: O(1).
   *
   *         <p>Start at top-right (or bottom-left). Move left if target smaller,
   *         move down if target larger.
   */
  public static boolean searchMatrixII(int[][] matrix, int target) {
    if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
      return false;
    }

    int row = 0;
    int col = matrix[0].length - 1;

    while (row < matrix.length && col >= 0) {
      if (matrix[row][col] == target) {
        return true;
      } else if (matrix[row][col] > target) {
        col--;
      } else {
        row++;
      }
    }

    return false;
  }

  /**
   * Find Peak Element (element greater than neighbors).
   *
   * @param nums input array
   * @return index of a peak element (element > neighbors)
   * @throws NullPointerException if nums is null
   *
   *         <p>LeetCode 162 (Medium). Time: O(log n).
   *         Space: O(1).
   *
   *         <p>Key: nums[-1] = nums[n] = -inf. Compare nums[mid] with
   *         nums[mid+1]: if nums[mid] < nums[mid+1], peak in right half;
   *         else in left half (including mid).
   */
  public static int findPeakElement(int[] nums) {
    if (nums == null || nums.length == 0) {
      return -1;
    }

    int left = 0;
    int right = nums.length - 1;

    while (left < right) {
      int mid = left + (right - left) / 2;

      if (nums[mid] < nums[mid + 1]) {
        // Peak in right half
        left = mid + 1;
      } else {
        // Peak in left half or at mid
        right = mid;
      }
    }

    return left;
  }

  /**
   * Main method demonstrating all binary search variants.
   */
  public static void main(String[] args) {
    System.out.println("=== BINARY SEARCH VARIANTS ===\n");

    // Standard search
    int[] arr1 = {-1, 0, 3, 5, 9, 12};
    System.out.println("Search: " + search(arr1, 9));

    // Search insert position
    int[] arr2 = {1, 3, 5, 6};
    System.out.println("Insert Position (target=5): " +
        searchInsertPosition(arr2, 5));

    // Search range
    int[] arr3 = {5, 7, 7, 8, 8, 10};
    System.out.println("Search Range (target=8): " +
        Arrays.toString(searchRange(arr3, 8)));

    // Rotated array search
    int[] arr4 = {4, 5, 6, 7, 0, 1, 2};
    System.out.println("Rotated Array Search (target=0): " +
        searchInRotatedArray(arr4, 0));

    // Find min in rotated
    int[] arr5 = {3, 4, 5, 1, 2};
    System.out.println("Find Min in Rotated: " +
        findMinInRotatedArray(arr5));

    // Koko eating bananas
    int[] piles = {1, 1, 1, 1};
    System.out.println("Koko Eating (h=4): " +
        kokoEatingBananas(piles, 4));

    // Split array
    int[] nums1 = {1, 2, 3, 4, 5};
    System.out.println("Split Array (k=2): " +
        splitArrayLargestSum(nums1, 2));

    // Ship packages
    int[] weights = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    System.out.println("Ship Packages (days=5): " +
        shipPackagesWithinDays(weights, 5));

    // Search matrix
    int[][] matrix1 = {{1, 3, 5, 7}, {10, 11, 16, 20}, {23, 30, 34, 60}};
    System.out.println("Search Matrix (target=3): " +
        searchMatrix(matrix1, 3));

    // Search matrix II
    int[][] matrix2 = {{1, 4, 7, 11, 15}, {2, 5, 8, 12, 19},
        {3, 6, 9, 16, 22}, {10, 13, 14, 17, 24}, {18, 21, 23, 26, 30}};
    System.out.println("Search Matrix II (target=13): " +
        searchMatrixII(matrix2, 13));

    // Find peak
    int[] arr6 = {1, 2, 3, 1};
    System.out.println("Find Peak Element: " +
        findPeakElement(arr6));
  }
}

File: C++ Playground/07_BinarySearch.cpp
/// Binary Search Playground
/// Invariant: arr[0..left-1] < target <= arr[left..n-1];
/// left + (right-left)/2 — NEVER (left+right)/2; left <= right for exact
/// match, left < right for left-boundary; answer-space: feasibility function
/// then binary search over [lo, hi]; Tricky: off-by-one in right = mid vs
/// right = mid-1.

#include <algorithm>
#include <iostream>
#include <vector>

namespace binary_search_playground {

constexpr int kNotFound = -1;

// ============================================================================
// STANDARD BINARY SEARCH
// ============================================================================

/// @brief Find exact target in sorted array.
/// @param nums Sorted array of integers
/// @param target Value to find
/// @return Index of target, or -1 if not found
/// @example Search({-1,0,3,5,9,12}, 9) -> 4
/// @constraints Array is sorted; return any occurrence if duplicates
/// @time O(log n) @space O(1)
[[nodiscard]] int Search(const std::vector<int>& nums,
                          int target) noexcept {
  int left = 0;
  int right = nums.size() - 1;

  while (left <= right) {
    int mid = left + (right - left) / 2;
    if (nums[mid] == target) {
      return mid;
    }
    if (nums[mid] < target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }
  return kNotFound;
}

/// @brief Find insert position to maintain sorted order.
/// @param nums Sorted array
/// @param target Value to insert
/// @return Index where target should be inserted
/// @example SearchInsertPosition({1,3,5,6}, 5) -> 2
/// @example SearchInsertPosition({1,3,5,6}, 7) -> 4
/// @constraints Invariant: result is where target belongs
/// @time O(log n) @space O(1)
[[nodiscard]] int SearchInsertPosition(const std::vector<int>& nums,
                                        int target) noexcept {
  int left = 0;
  int right = nums.size();

  while (left < right) {
    int mid = left + (right - left) / 2;
    if (nums[mid] < target) {
      left = mid + 1;
    } else {
      right = mid;
    }
  }
  return left;
}

// ============================================================================
// FIRST/LAST OCCURRENCE
// ============================================================================

/// @brief Find first (leftmost) occurrence of target.
/// @param nums Sorted array with possible duplicates
/// @param target Value to find
/// @return Index of first occurrence, or -1
/// @example FindFirstOccurrence({5,7,7,8,8,10}, 8) -> 3
/// @constraints Invariant: find leftmost equal value
/// @time O(log n) @space O(1)
[[nodiscard]] int FindFirstOccurrence(const std::vector<int>& nums,
                                       int target) noexcept {
  int left = 0;
  int right = nums.size();

  while (left < right) {
    int mid = left + (right - left) / 2;
    if (nums[mid] < target) {
      left = mid + 1;
    } else {
      right = mid;
    }
  }

  if (left < nums.size() && nums[left] == target) {
    return left;
  }
  return kNotFound;
}

/// @brief Find last (rightmost) occurrence of target.
/// @param nums Sorted array with possible duplicates
/// @param target Value to find
/// @return Index of last occurrence, or -1
/// @example FindLastOccurrence({5,7,7,8,8,10}, 8) -> 4
/// @constraints Invariant: find rightmost equal value
/// @time O(log n) @space O(1)
[[nodiscard]] int FindLastOccurrence(const std::vector<int>& nums,
                                      int target) noexcept {
  int left = 0;
  int right = nums.size();

  while (left < right) {
    int mid = left + (right - left) / 2;
    if (nums[mid] <= target) {
      left = mid + 1;
    } else {
      right = mid;
    }
  }

  if (left > 0 && nums[left - 1] == target) {
    return left - 1;
  }
  return kNotFound;
}

/// @brief Find first and last occurrence range.
/// @param nums Sorted array with possible duplicates
/// @param target Value to find
/// @return [first_idx, last_idx] or [-1, -1] if not found
/// @example SearchRange({5,7,7,8,8,10}, 8) -> [3,4]
/// @constraints Use first+last occurrence helpers
/// @time O(log n) @space O(1)
[[nodiscard]] std::vector<int>
SearchRange(const std::vector<int>& nums, int target) noexcept {
  int first = FindFirstOccurrence(nums, target);
  if (first == kNotFound) {
    return {-1, -1};
  }
  int last = FindLastOccurrence(nums, target);
  return {first, last};
}

// ============================================================================
// ROTATED SORTED ARRAY
// ============================================================================

/// @brief Search in rotated sorted array.
/// @param nums Rotated sorted array (no duplicates)
/// @param target Value to find
/// @return Index of target, or -1
/// @example SearchInRotatedArray({4,5,6,7,0,1,2}, 0) -> 4
/// @constraints Determine which half is sorted; search accordingly
/// @time O(log n) @space O(1)
/// Tricky: nums[left] <= nums[mid] means left half is sorted; else right
[[nodiscard]] int SearchInRotatedArray(const std::vector<int>& nums,
                                        int target) noexcept {
  int left = 0;
  int right = nums.size() - 1;

  while (left <= right) {
    int mid = left + (right - left) / 2;
    if (nums[mid] == target) {
      return mid;
    }

    if (nums[left] <= nums[mid]) {
      // Left half is sorted
      if (nums[left] <= target && target < nums[mid]) {
        right = mid - 1;
      } else {
        left = mid + 1;
      }
    } else {
      // Right half is sorted
      if (nums[mid] < target && target <= nums[right]) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }
  }
  return kNotFound;
}

/// @brief Find minimum element in rotated sorted array.
/// @param nums Rotated sorted array
/// @return Minimum value
/// @example FindMinInRotatedArray({3,4,5,1,2}) -> 1
/// @constraints nums[0] != nums[right] guaranteed (no duplicates)
/// @time O(log n) @space O(1)
[[nodiscard]] int FindMinInRotatedArray(const std::vector<int>& nums) noexcept {
  int left = 0;
  int right = nums.size() - 1;

  while (left < right) {
    int mid = left + (right - left) / 2;
    if (nums[mid] > nums[right]) {
      // Min is in right half
      left = mid + 1;
    } else {
      // Min is in left half or is mid
      right = mid;
    }
  }
  return nums[left];
}

// ============================================================================
// ANSWER SPACE (FEASIBILITY + BINARY SEARCH)
// ============================================================================

/// @brief Minimum eating speed for Koko to eat all bananas in h hours.
/// @param piles Pile sizes
/// @param h Hour limit
/// @return Minimum eating speed (bananas per hour)
/// @example KokoEatingBananas({1,1,1,1}, 4) -> 1
/// @example KokoEatingBananas({312884132}, 968709470) -> 1
/// @constraints Binary search on speed [1, max_pile]; feasibility check
/// @time O(n log(max_pile)) @space O(1)
static [[nodiscard]] bool CanEat(const std::vector<int>& piles,
                                  long long speed, int h) noexcept {
  long long hours = 0;
  for (int pile : piles) {
    hours += (pile + speed - 1) / speed;  // Ceiling division
    if (hours > h) return false;
  }
  return true;
}

[[nodiscard]] int KokoEatingBananas(const std::vector<int>& piles,
                                     int h) noexcept {
  int left = 1;
  int right = *std::max_element(piles.begin(), piles.end());

  while (left < right) {
    int mid = left + (right - left) / 2;
    if (CanEat(piles, mid, h)) {
      right = mid;
    } else {
      left = mid + 1;
    }
  }
  return left;
}

/// @brief Minimum largest sum when splitting array into k subarrays.
/// @param nums Array of integers
/// @param k Number of subarrays
/// @return Minimum of maximum subarray sums
/// @example SplitArrayLargestSum({7,2,5,10,8}, 2) -> 18
/// @constraints Binary search on answer [max_single, sum_all]
/// @time O(n log(sum)) @space O(1)
static [[nodiscard]] bool CanSplit(const std::vector<int>& nums, int k,
                                    long long max_sum) noexcept {
  int count = 1;
  long long current_sum = 0;
  for (int num : nums) {
    if (current_sum + num > max_sum) {
      ++count;
      current_sum = num;
      if (count > k) return false;
    } else {
      current_sum += num;
    }
  }
  return true;
}

[[nodiscard]] int SplitArrayLargestSum(const std::vector<int>& nums,
                                        int k) noexcept {
  long long left = *std::max_element(nums.begin(), nums.end());
  long long right = 0;
  for (int num : nums) {
    right += num;
  }

  while (left < right) {
    long long mid = left + (right - left) / 2;
    if (CanSplit(nums, k, mid)) {
      right = mid;
    } else {
      left = mid + 1;
    }
  }
  return left;
}

/// @brief Minimum days to ship all packages within deadline.
/// @param weights Package weights
/// @param days Day limit
/// @return Minimum shipping capacity (kg per day)
/// @example ShipPackages({1,2,3,4,5,6,7,8,9,10}, 5) -> 15
/// @constraints Binary search on capacity; greedy packing
/// @time O(n log(sum)) @space O(1)
[[nodiscard]] int ShipPackages(const std::vector<int>& weights,
                                int days) noexcept {
  long long left = *std::max_element(weights.begin(), weights.end());
  long long right = 0;
  for (int w : weights) {
    right += w;
  }

  while (left < right) {
    long long mid = left + (right - left) / 2;
    int required_days = 1;
    long long current_weight = 0;

    for (int w : weights) {
      if (current_weight + w > mid) {
        ++required_days;
        current_weight = w;
      } else {
        current_weight += w;
      }
    }

    if (required_days <= days) {
      right = mid;
    } else {
      left = mid + 1;
    }
  }
  return left;
}

// ============================================================================
// 2D MATRIX
// ============================================================================

/// @brief Search in 2D matrix (sorted row-wise and col-wise).
/// @param matrix m x n matrix, each row and column sorted
/// @param target Value to find
/// @return true if found, false otherwise
/// @example SearchMatrix({{1,3,5,7},{10,11,16,20},{23,30,34,60}}, 3) -> true
/// @constraints Flatten to 1D indices: row = mid/cols, col = mid%cols
/// @time O(log(m*n)) @space O(1)
[[nodiscard]] bool SearchMatrix(const std::vector<std::vector<int>>& matrix,
                                 int target) noexcept {
  if (matrix.empty() || matrix[0].empty()) return false;

  int m = matrix.size();
  int n = matrix[0].size();
  int left = 0;
  int right = m * n - 1;

  while (left <= right) {
    int mid = left + (right - left) / 2;
    int value = matrix[mid / n][mid % n];

    if (value == target) {
      return true;
    }
    if (value < target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }
  return false;
}

/// @brief Search in 2D matrix (sorted rows, but not columns).
/// @param matrix m x n matrix, each row sorted; leftmost of row > rightmost
/// of prev row? No
/// @param target Value to find
/// @return true if found, false otherwise
/// @example SearchMatrixII({{1,4,7,11,15},{2,5,8,12,19},...}, 5) -> true
/// @constraints Start from top-right; move left if too large, down if too
/// small
/// @time O(m+n) @space O(1)
/// Tricky: Top-right pointer strategy (could also use bottom-left)
[[nodiscard]] bool SearchMatrixII(const std::vector<std::vector<int>>& matrix,
                                   int target) noexcept {
  if (matrix.empty() || matrix[0].empty()) return false;

  int row = 0;
  int col = matrix[0].size() - 1;

  while (row < static_cast<int>(matrix.size()) && col >= 0) {
    if (matrix[row][col] == target) {
      return true;
    }
    if (matrix[row][col] > target) {
      --col;
    } else {
      ++row;
    }
  }
  return false;
}

// ============================================================================
// PEAK ELEMENT
// ============================================================================

/// @brief Find peak element (greater than neighbors).
/// @param nums Array where nums[i] != nums[i+1]
/// @return Index of any peak element
/// @example FindPeakElement({1,2,1,3,5,6,4}) -> 1 or 5
/// @constraints Array is finite; ends imply edges are < their single neighbor
/// @time O(log n) @space O(1)
/// Tricky: Compare nums[mid] with nums[mid+1]; if > go left, else go right
[[nodiscard]] int FindPeakElement(const std::vector<int>& nums) noexcept {
  int left = 0;
  int right = nums.size() - 1;

  while (left < right) {
    int mid = left + (right - left) / 2;
    if (nums[mid] > nums[mid + 1]) {
      right = mid;
    } else {
      left = mid + 1;
    }
  }
  return left;
}

}  // namespace binary_search_playground

// ============================================================================
// TESTS
// ============================================================================

int main() {
  using namespace binary_search_playground;

  std::cout << "=== Binary Search Playground ===\n\n";

  std::cout << "Search({-1,0,3,5,9,12}, 9): ";
  {
    auto result = Search({-1, 0, 3, 5, 9, 12}, 9);
    std::cout << result << "\n";
  }

  std::cout << "SearchInsertPosition({1,3,5,6}, 5): ";
  {
    auto result = SearchInsertPosition({1, 3, 5, 6}, 5);
    std::cout << result << "\n";
  }

  std::cout << "FindFirstOccurrence({5,7,7,8,8,10}, 8): ";
  {
    auto result = FindFirstOccurrence({5, 7, 7, 8, 8, 10}, 8);
    std::cout << result << "\n";
  }

  std::cout << "FindLastOccurrence({5,7,7,8,8,10}, 8): ";
  {
    auto result = FindLastOccurrence({5, 7, 7, 8, 8, 10}, 8);
    std::cout << result << "\n";
  }

  std::cout << "SearchRange({5,7,7,8,8,10}, 8): ";
  {
    auto result = SearchRange({5, 7, 7, 8, 8, 10}, 8);
    std::cout << "[" << result[0] << ", " << result[1] << "]\n";
  }

  std::cout << "SearchInRotatedArray({4,5,6,7,0,1,2}, 0): ";
  {
    auto result = SearchInRotatedArray({4, 5, 6, 7, 0, 1, 2}, 0);
    std::cout << result << "\n";
  }

  std::cout << "FindMinInRotatedArray({3,4,5,1,2}): ";
  {
    auto result = FindMinInRotatedArray({3, 4, 5, 1, 2});
    std::cout << result << "\n";
  }

  std::cout << "KokoEatingBananas({1,1,1,1}, 4): ";
  {
    auto result = KokoEatingBananas({1, 1, 1, 1}, 4);
    std::cout << result << "\n";
  }

  std::cout << "SplitArrayLargestSum({7,2,5,10,8}, 2): ";
  {
    auto result = SplitArrayLargestSum({7, 2, 5, 10, 8}, 2);
    std::cout << result << "\n";
  }

  std::cout << "ShipPackages({1,2,3,4,5,6,7,8,9,10}, 5): ";
  {
    auto result = ShipPackages({1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 5);
    std::cout << result << "\n";
  }

  std::cout << "SearchMatrix({...}, 3): ";
  {
    auto result = SearchMatrix({{1, 3, 5, 7}, {10, 11, 16, 20}}, 3);
    std::cout << (result ? "true" : "false") << "\n";
  }

  std::cout << "FindPeakElement({1,2,1,3,5,6,4}): ";
  {
    auto result = FindPeakElement({1, 2, 1, 3, 5, 6, 4});
    std::cout << result << "\n";
  }

  return 0;
}
